ext/openssl/openssl.c

```diff
@@ -3566,13 +3610,43 @@ static EVP_PKEY *php_openssl_pkey_from_zval(zval *val, int public_key, char *pas
 				TMP_CLEAN;
 			}
 		}
+		if (Z_STRLEN_P(val) > 7 && memcmp(Z_STRVAL_P(val), "pkcs11:", sizeof("pkcs11:") - 1) == 0) {
+			char *verbose = NULL;
+			engine = ENGINE_by_id("pkcs11");
+			if (engine == NULL) {
+				php_error_docref(NULL, E_WARNING, "Cannot load PKCS11 engine");
+				TMP_CLEAN;
+			}
+			verbose = getenv("OPENSSL_ENGINE_VERBOSE");
+			if (verbose) {
+				if (!ENGINE_ctrl_cmd_string(engine, "VERBOSE", NULL, 0)) {
+					ENGINE_free(engine);
+					TMP_CLEAN;
+				}
+			}
+			if (!ENGINE_init(engine)) {
+				ENGINE_free(engine);
+				ENGINE_finish(engine);
```

Comment:
Is there any point to do finish if init failed?

Reply:
right, it should be removed. Thanks.

Reply:
done: https://github.com/php/php-src/pull/6860/commits/bc27cf25e817c97f77b162ecfe1806bc571dda28